File: shoring/flywheel/src/main/java/com/panopset/flywheel/ImpliedQuitFilter.java
package com.panopset.flywheel;

import java.util.ArrayList;
import java.util.List;

/**
 * Implied quit filter.
 * <p>
 * Adds implied Command Quits to the commands vector.
 * </p>
 */
class ImpliedQuitFilter {

  /**
   * Add implied quits.
   *
   * @param commands
   *          Commands to add implied quits to.
   * @return List of commands with implied Command Quits added.
   */
  List<Command> addImpliedQuits(final List<Command> commands) throws FlywheelException {
    List<Command> rtn = new ArrayList<Command>();
    if (commands == null) {
      return rtn;
    }
    Command first = null;
    boolean checkOnce = false;
    for (Command command : commands) {
      if (first == null) {
        first = command;
      }
      if (command instanceof CommandFile) {
        insertImpliedQuitBefore(command);
        command.getTemplate().setCommandFile((CommandFile) command);
      }
      if (command.getNext() == null
          && command.getTemplate().getCommandFile() != null) {
        if (checkOnce) {
          throw new FlywheelException("Structure error.");
        }
        ImpliedQuit iq = new ImpliedQuit(command.getTemplate());
        command.setNext(iq);
        iq.setPrev(command);
        checkOnce = true;
      }
    }
    Command command = first;
    while (command != null) {
      rtn.add(command);
      command = command.getNext();
    }
    return rtn;
  }

  /**
   * Insert implied quit.
   * 
   * @param command
   *          Command to insert before.
   */
  private void insertImpliedQuitBefore(final Command command) {
    if (command.getTemplate().getCommandFile() == null) {
      return;
    }
    ImpliedQuit iq = new ImpliedQuit(command.getTemplate());
    iq.setNext(command);
    if (command.getPrev() != null) {
      iq.setPrev(command.getPrev());
      command.getPrev().setNext(iq);
    }
  }

}

File: code/shoring/flywheel/src/main/kotlin/com/panopset/flywheel/ImpliedQuitFilter.kt
package com.panopset.flywheel

/**
 * Implied quit filter.
 *
 *
 * Adds implied Command Quits to the commands vector.
 *
 */
internal class ImpliedQuitFilter {
    /**
     * Add implied quits.
     *
     * @param commands
     * Commands to add implied quits to.
     * @return List of commands with implied Command Quits added.
     */
    fun addImpliedQuits(commands: List<Command>): List<Command> {
        val rtn: MutableList<Command> = ArrayList()
        var first: Command? = null
        var checkOnce = false
        for (command in commands) {
            if (first == null) {
                first = command
            }
            if (command is CommandFile) {
                insertImpliedQuitBefore(command)
                command.template.commandFile = command
            }
            if (command.next == null
                && command.template.commandFile != null
            ) {
                if (checkOnce) {
                    throw FlywheelException("Structure error.")
                }
                val iq = ImpliedQuit(command.template)
                command.next = iq
                iq.prev = command
                checkOnce = true
            }
        }
        var command = first
        while (command != null) {
            rtn.add(command)
            command = command.next
        }
        return rtn
    }

    /**
     * Insert implied quit.
     *
     * @param command
     * Command to insert before.
     */
    private fun insertImpliedQuitBefore(command: Command) {
        if (command.template.commandFile == null) {
            return
        }
        val iq = ImpliedQuit(command.template)
        iq.next = command
        if (command.prev != null) {
            iq.prev = command.prev
            command.prev!!.next = iq
        }
    }
}
